core/lib/zksync_core/src/api_server/web3/backend_jsonrpsee/batch_limiter_middleware.rs

```diff
@@ -0,0 +1,93 @@
+use std::num::NonZeroU32;
+
+use governor::{
+    clock::DefaultClock,
+    middleware::NoOpMiddleware,
+    state::{InMemoryState, NotKeyed},
+    Quota, RateLimiter,
+};
+use vise::{
+    Buckets, Counter, EncodeLabelSet, EncodeLabelValue, Family, GaugeGuard, Histogram, Metrics,
+};
+use zksync_web3_decl::jsonrpsee::{
+    server::middleware::rpc::{layer::ResponseFuture, RpcServiceT},
+    types::{error::ErrorCode, ErrorObject, Request},
+    MethodResponse,
+};
+
+use crate::api_server::web3::metrics::API_METRICS;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, EncodeLabelValue, EncodeLabelSet)]
+#[metrics(label = "transport", rename_all = "snake_case")]
+pub(crate) enum Transport {
+    Ws,
+}
+
+#[derive(Debug, Metrics)]
+#[metrics(prefix = "api_jsonrpc_backend_batch")]
+struct LimitMiddlewareMetrics {
+    /// Number of rate-limited requests.
+    rate_limited: Family<Transport, Counter>,
+    /// Size of batch requests.
+    #[metrics(buckets = Buckets::exponential(1.0..=512.0, 2.0))]
+    size: Family<Transport, Histogram<usize>>,
+    /// Number of requests rejected by the limiter.
+    rejected: Family<Transport, Counter>,
+}
+
+#[vise::register]
+static METRICS: vise::Global<LimitMiddlewareMetrics> = vise::Global::new();
+
+/// A rate-limiting middleware.
+///
+/// `jsonrpsee` will allocate the instance of this struct once per session.
+pub(crate) struct LimitMiddleware<S> {
+    inner: S,
+    rate_limiter: Option<RateLimiter<NotKeyed, InMemoryState, DefaultClock, NoOpMiddleware>>,
+    transport: Transport,
+    __guard: GaugeGuard,
```

Comment:
Nit: One underscore is enough (it indicates that `guard` is not read, just used for its `Drop` implementation).

Reply:
renamed